beacon_node/beacon_chain/src/validator_custody.rs

```diff
@@ -0,0 +1,433 @@
+                .clone(),
+            persisted_is_supernode: context.persisted_is_supernode,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::MainnetEthSpec;
+
+    use super::*;
+
+    type E = MainnetEthSpec;
+
+    #[test]
+    fn no_validators() {
+        // supernode without validators
+        let custody_context = CustodyContext::new(true);
+        let spec = E::default_spec();
+        assert_eq!(
+            custody_context.custody_group_count(&spec),
+            spec.number_of_custody_groups
+        );
+        assert_eq!(
+            custody_context.advertised_custody_group_count(&spec),
+            spec.number_of_custody_groups
+        );
+    }
+
+    #[test]
+    fn fullnode() {
+        // fullnode without validators
+        let custody_context = CustodyContext::new(false);
+        let spec = E::default_spec();
+        assert_eq!(
+            custody_context.custody_group_count(&spec),
+            spec.custody_requirement,
+            "head custody count should be minimum spec custody requirement"
+        );
+        assert_eq!(
+            custody_context.advertised_custody_group_count(&spec),
+            spec.custody_requirement,
+            "advertised custody count should be minimum spec custody requirement"
+        );
+
+        // add 1 validator
+        custody_context.register_validator::<E>(vec![(0, 32_000_000_000)], Slot::new(0), &spec);
+
+        assert_eq!(
+            custody_context.custody_group_count(&spec),
+            spec.validator_custody_requirement + spec.custody_requirement,
+            "head custody count should increase with 1 validator"
+        );
+
+        assert_eq!(
+            custody_context.advertised_custody_group_count(&spec),
+            spec.custody_requirement,
+            "advertised custody count should not change"
+        );
+
+        // add 7 more validators to reach `validator_custody_requirement`
+        custody_context.register_validator::<E>(
+            vec![
+                (1, 32_000_000_000),
+                (2, 32_000_000_000),
+                (3, 32_000_000_000),
+                (4, 32_000_000_000),
+                (5, 32_000_000_000),
+                (6, 32_000_000_000),
+                (7, 32_000_000_000),
+            ],
+            Slot::new(0),
+            &spec,
+        );
+
+        assert_eq!(
+            custody_context.custody_group_count(&spec),
+            spec.validator_custody_requirement + spec.custody_requirement,
+            "head custody count should should be same as with 1 validator"
+        );
+
+        assert_eq!(
+            custody_context.advertised_custody_group_count(&spec),
+            spec.custody_requirement,
+            "advertised custody count should not change"
+        );
+
+        // adding 1 more validator should increase the custody count
+        custody_context.register_validator::<E>(vec![(8, 32_000_000_000)], Slot::new(0), &spec);
```

Comment:
shouldn't the custdoy group count be `spec.validator_custody_requirement` until the stake reaches `8 * eth per custody increment`?